src/toil/common.py

```diff
@@ -452,37 +452,47 @@ def __init__(self, options):
         super(Toil, self).__init__()
         self.options = options
         self.config = None
-        self.jobStore = None
-        self.batchSystem = None
-        self.jobCache = dict()
+        self._jobStore = None
+        self._batchSystem = None
+        self._jobCache = dict()
+        self._inContextManager = False
 
     def __enter__(self):
         """
         Derive configuration from the command line options, load the job store and, on restart,
         consolidate the derived configuration with the one from the previous invocation of the
         workflow.
         """
-        self.config = Config()
-        self.config.setOptions(self.options)
-        self.jobStore = self.loadOrCreateJobStore(self.config.jobStore,
-                                                  config=None if self.config.restart else self.config)
-        if self.config.restart:
-            # Reload configuration from job store
-            self.config = self.jobStore.config
+        try:
+            self.config = Config()
             self.config.setOptions(self.options)
-            self.jobStore.writeConfigToStore()
-
-        return self
+            self._jobStore = self.loadOrCreateJobStore(self.config.jobStore,
+                                                       config=None if self.config.restart else self.config)
+            if self.config.restart:
+                # Reload configuration from job store
+                self.config = self._jobStore.config
+                self.config.setOptions(self.options)
+                self._jobStore.writeConfigToStore()
+
+            return self
+        finally:
+            self._inContextManager = True
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         """
         Clean up after a workflow invocaton. Depending on the configuration, delete the job store.
         """
-        if (exc_type is not None and self.config.clean == "onError" or
-            exc_type is None and self.config.clean == "onSuccess" or
-            self.config.clean == "always"):
-            self.jobStore.deleteJobStore()
-        return False  # let exceptions through
+        try:
+            if (exc_type is not None and self.config.clean == "onError" or
+                exc_type is None and self.config.clean == "onSuccess" or
+                                     self.config.clean == "always"):
```

Comment:
Using reformat code in PyCharm gave me this (which I'm assuming is wrong)
```
 if (exc_type is not None and self.config.clean == "onError" or
                            exc_type is None and self.config.clean == "onSuccess" or
                        self.config.clean == "always"):
```

Is this ok?
```
    if (exc_type is not None and self.config.clean == "onError" or
        exc_type is None and self.config.clean == "onSuccess" or
         self.config.clean == "always"):
```